pkg/restmapper/caching.go

```diff
@@ -0,0 +1,115 @@
+package restmapper
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// cache is our cache of schema information.
+type cache struct {
+	mutex         sync.Mutex
+	groupVersions map[schema.GroupVersion]*cachedGroupVersion
+}
+
+// newCache is the constructor for a cache.
+func newCache() *cache {
+	return &cache{
+		groupVersions: make(map[schema.GroupVersion]*cachedGroupVersion),
+	}
+}
+
+// cachedGroupVersion caches (all) the resource information for a particular groupversion.
+type cachedGroupVersion struct {
+	gv    schema.GroupVersion
+	mutex sync.Mutex
+	kinds map[string]cachedGVR
+}
+
+// cachedGVR caches the information for a particular resource.
+type cachedGVR struct {
+	Resource string
+	Scope    meta.RESTScope
+}
+
+// findRESTMapping returns the RESTMapping for the specified GVK, querying discovery if not cached.
+func (c *cache) findRESTMapping(discovery discovery.DiscoveryInterface, gv schema.GroupVersion, kind string) (*meta.RESTMapping, error) {
+	c.mutex.Lock()
+	cached := c.groupVersions[gv]
+	if cached == nil {
+		cached = &cachedGroupVersion{gv: gv}
+		c.groupVersions[gv] = cached
+	}
+	c.mutex.Unlock()
+	return cached.findRESTMapping(discovery, kind)
+}
+
+// findRESTMapping returns the RESTMapping for the specified GVK, querying discovery if not cached.
+func (c *cachedGroupVersion) findRESTMapping(discovery discovery.DiscoveryInterface, kind string) (*meta.RESTMapping, error) {
+	kinds, err := c.fetch(discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	cached, found := kinds[kind]
+	if !found {
+		return nil, nil
+	}
+	return &meta.RESTMapping{
+		Resource:         c.gv.WithResource(cached.Resource),
+		GroupVersionKind: c.gv.WithKind(kind),
+		Scope:            cached.Scope,
+	}, nil
+}
+
+// fetch returns the metadata, fetching it if not cached.
+func (c *cachedGroupVersion) fetch(discovery discovery.DiscoveryInterface) (map[string]cachedGVR, error) {
```

Comment:
I'm not sure I understand how this is different from the existing restmapper fundamentally? Except that you already have the groupversion so you don't need to target the `/api/` discovery root endpoint?

Reply:
So the current restmapper does discovery on every GV in the system.  So when we register more CRDs, every controller slows down.  Most controllers only work with a small subset of GVs, and so the goal is that they only load the ones they care about.  There's a small memory win, but it's really about much faster startup time (and avoiding the "thundering herd" effect).